### DISCUSSION TITLE
API called in the wrong state.

### DISCUSSION THREAD

**Discussion by kshirsagardeepika** [reach-sh/reach-lang#917] [Q&A] [answered]
```
export const main = Reach.App(() => {
  const A = Participant('Creator', {
    deployed: Fun([], Null),
    setAmount: UInt,
    setToken: Token,
  });
  const B = API('Buyer', {
    payMe:  Fun([], Null),
    myAddr: Fun([Address], Null),
  });
  init();

  A.publish();
  commit();
  A.only(() => {
    const [amt, tok]= declassify([interact.setAmount, interact.setToken]);
  });
  A.publish(amt, tok);
  A.interact.deployed();
  
  var amount = amt;
  invariant( balance() == 0 && balance(tok) >= 0 );
  
  while(true) {
    commit();
    A.pay([[1, tok]]);
    commit();
    // paySpec([tok]);
    const [[me], k1] = 
      call(B.myAddr).assume((a)=> {
        assume(a == this, 'No address'); })
        k1(null);
    commit();
    const [[], k2] =
      call(B.payMe)
          .pay(() => amount);
    k2(null);
    transfer(amount).to(A);
    transfer(1, tok).to(me);
    continue;
  }
  commit();
});

```

Getting error as 
**Error: Buyer.myAddr errored with Error: Assertion failed: API called in the wrong state. Currently in state: 6, expected:  [5]**
while calling api from frontend.
 reach version - 0.1.10
reach hashes - 84dc282c
reach-sh/stdlib - 0.1.10-rc.4

**Comment by jeapostrophe** [comment, marked as answer]
You are trying to call an API function when it is not allowed. My guess is that you are trying to call `payMe` when the program is waiting for a call to `myAddr`

**Reply by kshirsagardeepika** [reply]
What could be solution to notify user to wait until previous function finish? Can you provide some sample code?

**Reply by jeapostrophe** [reply]
My advice is to add an Event and only allow the UI to show the option to call that function when the event has triggered

**Reply by kshirsagardeepika** [reply]
As suggested added event as
```  
while(true) {
    commit();
    A.pay([[1, tok]]);
    E.announce('code0')
    commit();
    const [[me], k1] = 
      call(B.myAddr).assume((a)=> {
        assume(a == this, 'No address'); })
        k1(null);
    E.announce('code1')
    commit();
    const [[], k2] =
      call(B.payMe)
          .pay(() => amount);
          k2(null);
    E.announce('code2')  
    transfer(amount).to(A);
    transfer(1, tok).to(me);
    E.announce('code3')
    continue;
  }
```

and noticed that in FE it everytime logs only code0. If try with different buyer again getting same error as 
**Buyer.myAddr errored with Error: Assertion failed: API called in the wrong state. Currently in state: 4, expected: [5]**

**Reply by jeapostrophe** [reply]
Are you waiting for three announces? You need to wait for the event that happens right before the API call is allowed to be call

**Reply by kshirsagardeepika** [reply]
Updated code as
```const [[me], k1] = 
      call(B.myAddr).assume((a)=> {
        assume(a == this, 'No address'); })
        E.announce('code1')
        k1(null);
```
but announce never triggered in FE and nothing happen. 
If address is received then only I want to have pay and transfer to execute, am I doing something wrong?

**Reply by jeapostrophe** [reply]
I can't understand what you're writing.

In the code that you've written, you will only see `announce('code1')` if `B.myAddr()` is called.

**Reply by kshirsagardeepika** [reply]
complete code snippet
```
while(true) {
    commit();
    A.pay([[1, tok]]);
    commit();
    const [[me], k1] = 
      call(B.myAddr).assume((a)=> {
        assume(a == this, 'No address'); })
        k1(null);
    E.announce('code1');
    commit();
    const [[], k2] =
      call(B.payMe)
          .pay(() => amount);
          k2(null);
   
    transfer(amount).to(A);
    transfer(1, tok).to(me);
    continue;
  }
```

But  `announce('code1')` is **not displayed** when B.myAddr() is called. Can you please explain why announce is never called?
If I write E.announce() before calling api only that event display in FE.

**Reply by jeapostrophe** [reply]
I didn't say "displayed". I said "see". I mean that the event will only be generated if you call it. I have no idea what your frontend looks like or if you have it set up to actually receive the event.

**Reply by kshirsagardeepika** [reply]
In FE 
`let eve = this.attachctc.e.Announcer.announce.next();
console.log(eve);`
It never display in log.

**Reply by jeapostrophe** [reply]
You are not waiting for the log. `next` returns a promise

**Reply by kshirsagardeepika** [reply]
I am not able to figure out the issue so far. My requirement is, Creator sets token/nft and its price. Idea is multiple buyer should able to buy the NFTmicro . For this initially Creator is setting NFT details and  in while loop Buyer is calling API to set his address and amount of nft  and then Creator is paying for the token once address is set. Then finally when buyer clicks buy button in the front-end I am calling PayMe API.

```
export const main = Reach.App(() => {
const A = Participant('Creator', {
deployed: Fun([], Null),
setAmount: UInt,
setToken: Token,
});
const B = API('Buyer', {
payMe: Fun([], Bool),
myAddr: Fun([Address, UInt], Bool)
});
const E = Events('Announcer', {
announce: [Bytes(5)]
});
init();
A.publish();
commit();
A.interact.deployed();
A.only(() => {
const [amt, tok]= declassify([interact.setAmount, interact.setToken]);
});
A.publish(amt, tok);
var amount = amt;
invariant( balance() == 0 && balance(tok) >= 0 );
while(true) {
commit();
const [[me, count], k1] =
call(B.myAddr).assume((a, c)=> {
assume(a == this, 'No address');
check(c > 0, 'Count must be greater than 0');
})
k1(true);
commit();
A.pay([[count, tok]]);
E.announce('code1');
commit();
const [[], k2] =
call(B.payMe)
.pay(() => amount);
transfer(muldiv(balance(), 80, 100)).to(A);
transfer(balance()).to(A);
transfer(count, tok).to(me);
k2(true);
continue;
}
commit();
});
```

As suggested by @Jay I have added Event which is triggered after Creator pays for the token and then only Buyer can execute buy operation. It works fine. But what strange things I noticed in the  Event monitor in the front-end

this.attachctc.e.Announcer.announce.monitor((eve: any) => {
  console.log('monitor',eve);
});

Now, if there are 5 buyers it triggers 5 times but problem is if any user clicks buy(calls PayMe API) before finishing all the 5 events it breaks and gives error something like
“currently in state 5 expected in state 4”.

How can we solve this issue. In reality there could be 100 buyers and it is not logical to wait for all the 100 events before enabling buy.

**Reply by jeapostrophe** [reply]
I think you've written your program in a very strange way. You have B call `myAddr`, then A pays, then B comes back and calls `payMe`. Thus, every time through the while loop you have three steps that must be done in sequence. I think you really want just one call where a buyer buys with no interaction from A and no second interaction.

**Reply by kshirsagardeepika** [reply]
Thanks Jay.

I have modified contract as 
```
export const main = Reach.App(() => {
  const A = Participant('Creator', {
    deployed: Fun([], Null),
    setAmount: UInt,
    setToken: Token
  });
  const B = API('Buyer', {
    payMe:  Fun([], Bool),
    myAddr: Fun([Address, UInt], Bool)
  })
  init();
  A.publish();
  commit();
  A.interact.deployed();
  A.only(() => {
    const [amt, tok]= declassify([interact.setAmount, interact.setToken]);
  });
  A.publish(amt, tok);
  commit();
  A.pay([[10, tok]]);
  
  var [amount, nftcounter] = [amt, 10];
  invariant( balance() == 0 && balance(tok) == nftcounter );
  while(nftcounter > 0) {
    commit();
    // paySpec([tok]);
    const [[me, count], k1] =   
      call(B.myAddr)
        .assume((a, c)=> {
          assume(a == this, 'No address');
          check(c > 0, 'Count must be greater than 0');
        })
      k1(true);
        
    commit();
    const total = amount * count;
    const [[], k2] =  
      call(B.payMe)
        .pay(() => total);
        transfer(total).to(A);
        transfer(count, tok).to(me);
      k2(true);
    nftcounter = nftcounter - count;
    continue;
  }
  commit();
});
```
Now A will pay 10 tokens initially and multiple buyers can buy by specifying the count how much they want in `B.myAdd`. And when all tokens has been taken by  different users while loop should terminate.
But getting theorem failed error when compiled.
Can you please suggest what's wrong with this code?

**Reply by kshirsagardeepika** [reply]
This issue has been resolved by adding condition before transfer as below.
```
const [[], k2] = 
      call(B.payMe)
        .pay(() => total)
      
        if(num <= nftcounter) {
          transfer(total).to(A);
          transfer(num, tok).to(me);
        }
      k2(num);
```

Thanks.
